pkg/actions/tools/nix/attr.go

```diff
@@ -0,0 +1,82 @@
+package nix
+
+import (
+	_ "embed"
+	"encoding/json"
+	"fmt"
+	"path"
+	"strings"
+
+	"github.com/rsteube/carapace"
+	"github.com/spf13/cobra"
+)
+
+//go:embed attrComplete.nix
+var attrCompleteScript string
+
+func ActionAttr(cmd *cobra.Command, getSource func(*cobra.Command, carapace.Context) string) carapace.Action {
+	return carapace.ActionCallback(func(c carapace.Context) carapace.Action {
+		attrPath := c.CallbackValue
+		nixPath := ""
+		if cmd.Flag("include").Changed {
+			nixPath = cmd.Flag("include").Value.String()
+		}
+
+		source := getSource(cmd, c)
+
+		if strings.HasPrefix(source, "<") {
+			// expression is a local channel, use as-is
+		} else if strings.HasPrefix(source, "http:") || strings.HasPrefix(source, "https:") {
```

Comment:
Fetching the tarball might not be such a good idea here.  For example`nixpkgs-master.tar.gz` has `33MB`. Maybe only complete this for lokal channels/files?

Reply:
`nix` does cache these so it'd be only for the initial request. I think the zsh completions have a check to only complete if it's cached which I can maybe include here.

Reply:
I went and checked and I think the zsh completion logic does not work with channels/urls in the current version of nix. The way tarballs are cached has changed. Channels added via `nix-channel` can be checked in `~/.nix-defexpr/channels{,_root}`. Other urls can be checked in `~/.cache/nix/fetcher-cache-v1.sqlite`.

That said, anything you import (including local files) could do a separate import themselves and since nix is lazy, just because something is cached doesn't mean you won't evaluate something that needs to fetch something else later.

So I don't think we can reliably detect whether or not a particular source will cause a fetch. I think it would be best to just leave this as-is and just have the user cancel the completion if it's taking too long (or add some sort of timeout?)

Reply:
Yeah, a timeout would be good. There is currently no indication that something is happening (like a spinner) so for a user it is not clear what is blocking the terminal.
Think this should be done in carapace as general solution though, like `Action().Timeout(10*time.Seconds)`.
So let's just leave it like this for now.